### ISSUE TITLE
Support sync batch norm as a plugin

### ISSUE THREAD

**Issue by ananthsub** [PyTorchLightning/pytorch-lightning#4349]
## 🚀 Feature
Support sync batch norm as a plugin so we don't duplicate code across accelerators

### Motivation

It'll look very similar to this: https://github.com/PyTorchLightning/pytorch-lightning/pull/4285

### Pitch

```
import torch
from pytorch_lightning.core.lightning import LightningModule


class SyncBatchNormPlugin(object):
    """
    Plugin to link a custom sync batchnorm implementation to any arbitrary accelerator.
    Example::
        class MySyncBN(SyncBatchNormPlugin):

            def configure_sync_batchnorm(self, model):
                model = MySyncBNWrapper(model)
                return model
        my_sync_bn = MySyncBN()
        trainer = Trainer(sync_batchnorm=True, plugins=[my_sync_bn])
    """

    def configure_sync_batchnorm(self, model: LightningModule) -> LightningModule:
        """
        By default: adds global batchnorm for a model spread across multiple GPUs and nodes.
        Override this to synchronize batchnorm between specific process groups instead
        of the whole world or use a different sync_bn like `apex`'s version.
        Args:
            model: pointer to current :class:`LightningModule`.
        Return:
            LightningModule with batchnorm layers synchronized between process groups
        """
        model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model, process_group=None)

        return model
```

Do we even need the `sync_batchnorm` flag on the `Trainer` with this plugin?

**Comment by awaelchli** [comment]
> Do we even need the sync_batchnorm flag on the Trainer with this plugin?

This reminds me of the checkpoint_callback + callbacks list situation for ModelCheckpoint. The difference here is that there can be only one plugin of this type.

**Comment by ananthsub** [comment]
> This reminds me of the checkpoint_callback + callbacks list situation for ModelCheckpoint. The difference here is that there can be only one plugin of this type.

If we have the flag, we can use `sync_batchnorm=True` and get the default implementation of the plugin without creating and passing it to the trainer.

If we drop the flag, then to use `sync_batchnorm` you must create and pass a plugin (of any type that extends `SyncBatchNormPlugin`), and lightning would provide the default implementation for you to create and pass

**Comment by edenlightning** [comment]
@awaelchli @justusschock how does this fit in with the new API?

**Comment by justusschock** [comment]
I think it's not that easy to get it to the new API. We kind of would have to introduce a separate plugin class for this and define when and how this one should be called, since this is neither a precision, nor a training type plugin.

I'll sync with @awaelchi on this :)

**Comment by carmocca** [comment]
@awaelchli @ananthsub Should we shoot to have this for 1.6 if we want to mark the strategies as stable?

**Comment by awaelchli** [comment]
Yes, that sounds reasonable to me. We would follow the same pattern as for e.g. the checkpoint io plugin. This should fit in nicely.
